# Flight Operations Data Analysis

**Tools Used:** Python, Pandas  
**Dataset:** Flight Operations Dataset

This notebook performs a complete data analysis workflow using Pandas, including data inspection, selection, filtering, sorting, grouping, aggregation, data transformation, and date-time analysis.

In [ ]:
import pandas as pd

In [ ]:
df=pd.read_csv('/content/Day10_Flight_Operations_Dataset.csv')

## 2. Understanding the Dataset

In [ ]:
df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5


In [ ]:
df.shape

(180, 17)

In [ ]:
df.columns

Index(['Flight_ID', 'Flight_Date', 'Airline', 'Origin', 'Destination',
       'Aircraft', 'Travel_Class', 'Passengers', 'Seat_Capacity',
       'Average_Ticket_Price', 'Delay_Minutes', 'Flight_Status', 'Weather',
       'Booking_Channel', 'Avg_Baggage_Kg', 'Meal_Preference',
       'Passenger_Satisfaction'],
      dtype='object')

In [ ]:
df.tail()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
175,FL0176,2026-04-10,IndiGo,Ahmedabad,Bengaluru,Airbus A320,Economy,48,189,4298,0.0,On Time,Storm,Mobile App,11,Vegetarian,5
176,FL0177,2026-02-17,Vistara,Chennai,Mumbai,Airbus A320,Business,124,180,17306,20.0,Delayed,Cloudy,Mobile App,8,No Meal,5
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3
178,FL0179,2026-05-08,Vistara,Bengaluru,Kochi,Airbus A320,Economy,134,189,5413,20.0,Delayed,Rain,Online Travel Portal,9,No Meal,2
179,FL0180,2026-01-22,Akasa Air,Delhi,Mumbai,Airbus A321,Economy,144,210,3767,75.0,Delayed,Fog,Online Travel Portal,17,No Meal,5


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Flight_ID               180 non-null    object 
 1   Flight_Date             180 non-null    object 
 2   Airline                 180 non-null    object 
 3   Origin                  180 non-null    object 
 4   Destination             180 non-null    object 
 5   Aircraft                180 non-null    object 
 6   Travel_Class            180 non-null    object 
 7   Passengers              180 non-null    int64  
 8   Seat_Capacity           180 non-null    int64  
 9   Average_Ticket_Price    180 non-null    int64  
 10  Delay_Minutes           173 non-null    float64
 11  Flight_Status           180 non-null    object 
 12  Weather                 180 non-null    object 
 13  Booking_Channel         180 non-null    object 
 14  Avg_Baggage_Kg          180 non-null    in

In [ ]:
df.describe()

,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Avg_Baggage_Kg,Passenger_Satisfaction
count,180.000000,180.000000,180.000000,173.000000,180.000000,180.000000
mean,122.772222,197.238889,6111.677778,19.335260,15.150000,3.494444
std,45.635651,15.408176,3615.220040,29.029712,5.705506,1.169908
min,45.000000,180.000000,1949.000000,0.000000,5.000000,2.000000
25%,78.750000,186.000000,3815.500000,0.000000,10.000000,2.000000
50%,125.000000,189.000000,5380.000000,5.000000,15.500000,3.000000
75%,162.250000,210.000000,6790.000000,20.000000,20.000000,5.000000
max,210.000000,220.000000,27832.000000,110.000000,25.000000,5.000000


In [ ]:
df.isnull().sum()

,0
Flight_ID,0
Flight_Date,0
Airline,0
Origin,0
Destination,0
Aircraft,0
Travel_Class,0
Passengers,0
Seat_Capacity,0
Average_Ticket_Price,0


In [ ]:
df.duplicated().sum()

np.int64(0)

In [ ]:
df['Airline']

,Airline
0,SpiceJet
1,Akasa Air
2,Air India
3,Akasa Air
4,Air India
...,...
175,IndiGo
176,Vistara
177,IndiGo
178,Vistara


In [ ]:
df[['Airline', 'Origin', 'Destination']]

,Airline,Origin,Destination
0,SpiceJet,Mumbai,Bengaluru
1,Akasa Air,Delhi,Mumbai
2,Air India,Kochi,Delhi
3,Akasa Air,Srinagar,Delhi
4,Air India,Pune,Delhi
...,...,...,...
175,IndiGo,Ahmedabad,Bengaluru
176,Vistara,Chennai,Mumbai
177,IndiGo,Delhi,Mumbai
178,Vistara,Bengaluru,Kochi


In [ ]:
df[['Airline', 'Origin', 'Destination']].head(10)

,Airline,Origin,Destination
0,SpiceJet,Mumbai,Bengaluru
1,Akasa Air,Delhi,Mumbai
2,Air India,Kochi,Delhi
3,Akasa Air,Srinagar,Delhi
4,Air India,Pune,Delhi
5,Akasa Air,Kochi,Delhi
6,Air India Express,Delhi,Srinagar
7,Air India,Delhi,Chennai
8,Akasa Air,Delhi,Chennai
9,Akasa Air,Ahmedabad,Bengaluru


In [ ]:
df['Origin'].unique()

array(['Mumbai', 'Delhi', 'Kochi', 'Srinagar', 'Pune', 'Ahmedabad',
       'Lucknow', 'Jaipur', 'Bengaluru', 'Chennai', 'Hyderabad'],
      dtype=object)

In [ ]:
df[df['Origin'] == 'Kochi']

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
5,FL0006,2026-05-20,Akasa Air,Kochi,Delhi,Airbus A319,Economy,123,220,5636,35.0,Delayed,Cloudy,Travel Agency,13,Vegetarian,2
21,FL0022,2026-02-01,Vistara,Kochi,Delhi,Airbus A321,Economy,126,220,5887,75.0,Delayed,Clear,Airline Website,10,Vegetarian,2
39,FL0040,2026-03-21,Vistara,Kochi,Delhi,Airbus A319,Economy,178,220,5361,12.0,On Time,Cloudy,Mobile App,20,Vegan,4
48,FL0049,2026-04-19,SpiceJet,Kochi,Delhi,Airbus A319,Economy,98,189,4968,12.0,On Time,Cloudy,Mobile App,8,Vegetarian,4
49,FL0050,2026-06-23,Akasa Air,Kochi,Delhi,Airbus A320,Economy,74,186,3754,0.0,On Time,Cloudy,Travel Agency,14,Vegan,5
64,FL0065,2026-04-28,Air India Express,Kochi,Delhi,Airbus A319,Economy,165,186,4002,12.0,On Time,Clear,Mobile App,5,Vegan,3
68,FL0069,2026-03-09,SpiceJet,Kochi,Delhi,Airbus A321,Economy,107,186,5911,NaN,Cancelled,Cloudy,Online Travel Portal,23,Non-Vegetarian,5
76,FL0077,2026-02-27,Vistara,Kochi,Delhi,Airbus A319,Economy,150,180,3399,12.0,On Time,Storm,Travel Agency,12,Non-Vegetarian,5
96,FL0097,2026-03-10,Vistara,Kochi,Delhi,Airbus A319,Economy,90,186,6145,20.0,Delayed,Fog,Airline Website,13,Vegan,2


## 5. Sorting Data

In [ ]:
df.select_dtypes(include='number').columns.tolist()

['Passengers',
 'Seat_Capacity',
 'Average_Ticket_Price',
 'Delay_Minutes',
 'Avg_Baggage_Kg',
 'Passenger_Satisfaction']

In [ ]:
df[df['Delay_Minutes'] > 30]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
5,FL0006,2026-05-20,Akasa Air,Kochi,Delhi,Airbus A319,Economy,123,220,5636,35.0,Delayed,Cloudy,Travel Agency,13,Vegetarian,2
8,FL0009,2026-01-25,Akasa Air,Delhi,Chennai,Airbus A320,Economy,82,210,5893,75.0,Delayed,Cloudy,Mobile App,8,Vegetarian,4
21,FL0022,2026-02-01,Vistara,Kochi,Delhi,Airbus A321,Economy,126,220,5887,75.0,Delayed,Clear,Airline Website,10,Vegetarian,2
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.0,Delayed,Storm,Airline Website,10,Vegan,3
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4
32,FL0033,2026-04-27,Akasa Air,Ahmedabad,Bengaluru,Boeing 737,Business,66,180,17625,35.0,Delayed,Cloudy,Mobile App,15,No Meal,4
42,FL0043,2026-06-19,Vistara,Delhi,Chennai,Airbus A319,Economy,64,220,4029,50.0,Delayed,Clear,Mobile App,5,No Meal,3


In [ ]:
df[(df['Delay_Minutes'] > 30) & (df['Passengers'] > 200)]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4


In [ ]:
(df['Delay_Minutes'] > 30).sum()

np.int64(40)

## 6. Grouping and Aggregation

In [ ]:
df.sort_values(by='Delay_Minutes')

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,0.0,On Time,Fog,Airline Website,25,Vegan,5
6,FL0007,2026-06-17,Air India Express,Delhi,Srinagar,Airbus A319,Premium Economy,178,210,9430,0.0,On Time,Storm,Online Travel Portal,18,Vegetarian,4
15,FL0016,2026-05-15,Akasa Air,Mumbai,Bengaluru,Boeing 737,Premium Economy,69,186,7860,0.0,On Time,Storm,Mobile App,5,Vegan,3
12,FL0013,2026-05-14,SpiceJet,Pune,Delhi,Airbus A319,Economy,162,220,4689,0.0,On Time,Cloudy,Mobile App,6,Vegetarian,5
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
40,FL0041,2026-03-23,Akasa Air,Delhi,Chennai,Airbus A321,Economy,170,189,4013,NaN,Cancelled,Fog,Travel Agency,13,No Meal,3
65,FL0066,2026-03-06,SpiceJet,Srinagar,Delhi,Airbus A320,Economy,83,210,4418,NaN,Cancelled,Cloudy,Travel Agency,19,Vegetarian,5
68,FL0069,2026-03-09,SpiceJet,Kochi,Delhi,Airbus A321,Economy,107,186,5911,NaN,Cancelled,Cloudy,Online Travel Portal,23,Non-Vegetarian,5
110,FL0111,2026-05-16,Vistara,Delhi,Mumbai,Airbus A321,Premium Economy,124,186,6711,NaN,Cancelled,Cloudy,Mobile App,19,Vegetarian,4


In [ ]:
df.sort_values(by='Delay_Minutes', ascending=False)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3
82,FL0083,2026-04-18,Air India Express,Lucknow,Delhi,Airbus A319,Economy,71,180,2484,110.0,Delayed,Cloudy,Online Travel Portal,22,Vegetarian,4
57,FL0058,2026-04-14,IndiGo,Hyderabad,Kolkata,Boeing 737,Premium Economy,151,210,6834,110.0,Delayed,Cloudy,Travel Agency,16,Vegetarian,4
50,FL0051,2026-05-03,Air India,Delhi,Mumbai,Airbus A321,Economy,135,210,5129,110.0,Delayed,Rain,Online Travel Portal,17,Vegetarian,4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
40,FL0041,2026-03-23,Akasa Air,Delhi,Chennai,Airbus A321,Economy,170,189,4013,NaN,Cancelled,Fog,Travel Agency,13,No Meal,3
65,FL0066,2026-03-06,SpiceJet,Srinagar,Delhi,Airbus A320,Economy,83,210,4418,NaN,Cancelled,Cloudy,Travel Agency,19,Vegetarian,5
68,FL0069,2026-03-09,SpiceJet,Kochi,Delhi,Airbus A321,Economy,107,186,5911,NaN,Cancelled,Cloudy,Online Travel Portal,23,Non-Vegetarian,5
110,FL0111,2026-05-16,Vistara,Delhi,Mumbai,Airbus A321,Premium Economy,124,186,6711,NaN,Cancelled,Cloudy,Mobile App,19,Vegetarian,4


In [ ]:
df.sort_values(by='Delay_Minutes', ascending=False).head(10)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3
82,FL0083,2026-04-18,Air India Express,Lucknow,Delhi,Airbus A319,Economy,71,180,2484,110.0,Delayed,Cloudy,Online Travel Portal,22,Vegetarian,4
57,FL0058,2026-04-14,IndiGo,Hyderabad,Kolkata,Boeing 737,Premium Economy,151,210,6834,110.0,Delayed,Cloudy,Travel Agency,16,Vegetarian,4
50,FL0051,2026-05-03,Air India,Delhi,Mumbai,Airbus A321,Economy,135,210,5129,110.0,Delayed,Rain,Online Travel Portal,17,Vegetarian,4
128,FL0129,2026-01-04,SpiceJet,Lucknow,Delhi,Boeing 737,Premium Economy,67,210,5657,110.0,Delayed,Cloudy,Online Travel Portal,21,Non-Vegetarian,2
146,FL0147,2026-02-18,IndiGo,Mumbai,Kolkata,Airbus A319,Premium Economy,115,210,9106,110.0,Delayed,Rain,Online Travel Portal,19,Vegetarian,5
169,FL0170,2026-06-25,IndiGo,Mumbai,Kolkata,Airbus A319,Economy,130,180,3938,110.0,Delayed,Rain,Online Travel Portal,17,No Meal,4
161,FL0162,2026-01-18,Air India,Hyderabad,Kolkata,Airbus A321,Economy,128,189,5263,110.0,Delayed,Rain,Mobile App,6,Non-Vegetarian,3
21,FL0022,2026-02-01,Vistara,Kochi,Delhi,Airbus A321,Economy,126,220,5887,75.0,Delayed,Clear,Airline Website,10,Vegetarian,2


In [ ]:
df.sort_values(by='Passenger_Satisfaction', ascending=False).head(10)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
13,FL0014,2026-06-14,Air India,Lucknow,Delhi,Airbus A321,Economy,78,220,2744,0.0,On Time,Cloudy,Online Travel Portal,13,No Meal,5
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,0.0,On Time,Fog,Airline Website,25,Vegan,5
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5
19,FL0020,2026-02-01,SpiceJet,Mumbai,Kolkata,Airbus A320,Economy,125,210,5715,12.0,On Time,Fog,Travel Agency,24,Non-Vegetarian,5
14,FL0015,2026-01-21,Akasa Air,Lucknow,Delhi,Boeing 737,Economy,133,186,3911,0.0,On Time,Cloudy,Travel Agency,21,No Meal,5
12,FL0013,2026-05-14,SpiceJet,Pune,Delhi,Airbus A319,Economy,162,220,4689,0.0,On Time,Cloudy,Mobile App,6,Vegetarian,5
47,FL0048,2026-01-08,Air India,Delhi,Chennai,Airbus A321,Economy,68,210,3697,20.0,Delayed,Fog,Mobile App,16,Non-Vegetarian,5
37,FL0038,2026-03-27,Air India,Delhi,Srinagar,Airbus A320,Economy,166,210,5972,NaN,Cancelled,Rain,Travel Agency,5,Non-Vegetarian,5
41,FL0042,2026-03-12,IndiGo,Delhi,Chennai,Boeing 737,Economy,122,186,5670,12.0,On Time,Storm,Airline Website,21,Vegan,5


In [ ]:
df.select_dtypes(include='object').columns.tolist()

['Flight_ID',
 'Flight_Date',
 'Airline',
 'Origin',
 'Destination',
 'Aircraft',
 'Travel_Class',
 'Flight_Status',
 'Weather',
 'Booking_Channel',
 'Meal_Preference']

In [ ]:
df.groupby('Airline').size()

,0
Airline,
Air India,24
Air India Express,24
Akasa Air,31
IndiGo,32
SpiceJet,32
Vistara,37


In [ ]:
df.groupby('Airline')['Delay_Minutes'].mean()

,Delay_Minutes
Airline,
Air India,21.260870
Air India Express,11.291667
Akasa Air,14.400000
IndiGo,25.031250
SpiceJet,24.137931
Vistara,18.628571


In [ ]:
df.groupby('Airline')['Passenger_Satisfaction'].mean()

,Passenger_Satisfaction
Airline,
Air India,3.750000
Air India Express,3.666667
Akasa Air,3.419355
IndiGo,3.343750
SpiceJet,3.531250
Vistara,3.378378


In [ ]:
df.groupby('Airline').agg({
    'Passengers': 'sum',
    'Delay_Minutes': 'mean',
    'Passenger_Satisfaction': 'mean'
})

,Passengers,Delay_Minutes,Passenger_Satisfaction
Airline,,,
Air India,3074,21.260870,3.750000
Air India Express,3027,11.291667,3.666667
Akasa Air,3427,14.400000,3.419355
IndiGo,4158,25.031250,3.343750
SpiceJet,3929,24.137931,3.531250
Vistara,4484,18.628571,3.378378


In [ ]:
df.groupby('Origin')['Passengers'].sum().sort_values(ascending=False)

,Passengers
Origin,
Delhi,5448
Mumbai,3616
Bengaluru,2725
Srinagar,1957
Kochi,1657
Pune,1564
Ahmedabad,1411
Lucknow,1137
Jaipur,962


## 7. Data Transformation

In [ ]:
df['Delay_Category'] = df['Delay_Minutes'].apply(
    lambda x: 'On Time' if x == 0
    else 'Minor Delay' if x <= 30
    else 'Major Delay'
)

In [ ]:
df[['Delay_Minutes', 'Delay_Category']].head(10)

,Delay_Minutes,Delay_Category
0,50.0,Major Delay
1,0.0,On Time
2,5.0,Minor Delay
3,35.0,Major Delay
4,20.0,Minor Delay
5,35.0,Major Delay
6,0.0,On Time
7,0.0,On Time
8,75.0,Major Delay
9,12.0,Minor Delay


In [ ]:
df['Seat_Utilization'] = (
    df['Passengers'] / df['Seat_Capacity'] * 100
)

In [ ]:
df[['Passengers', 'Seat_Capacity', 'Seat_Utilization']].head(10)

,Passengers,Seat_Capacity,Seat_Utilization
0,210,180,116.666667
1,70,210,33.333333
2,67,220,30.454545
3,153,220,69.545455
4,178,180,98.888889
5,123,220,55.909091
6,178,210,84.761905
7,157,220,71.363636
8,82,210,39.047619
9,125,220,56.818182


In [ ]:
df['Seat_Utilization'] = df['Seat_Utilization'].round(2)

In [ ]:
df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Delay_Category,Seat_Utilization
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,Major Delay,116.67
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5,On Time,33.33
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2,Minor Delay,30.45
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2,Major Delay,69.55
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5,Minor Delay,98.89


In [ ]:
df['Airline'].value_counts()

,count
Airline,
Vistara,37
SpiceJet,32
IndiGo,32
Akasa Air,31
Air India Express,24
Air India,24


In [ ]:
df.groupby('Airline')['Delay_Minutes'].mean().sort_values(ascending=False)

,Delay_Minutes
Airline,
IndiGo,25.031250
SpiceJet,24.137931
Air India,21.260870
Vistara,18.628571
Akasa Air,14.400000
Air India Express,11.291667


In [ ]:
df.groupby('Airline')['Passenger_Satisfaction'].mean().sort_values(ascending=False)

,Passenger_Satisfaction
Airline,
Air India,3.750000
Air India Express,3.666667
SpiceJet,3.531250
Akasa Air,3.419355
Vistara,3.378378
IndiGo,3.343750


In [ ]:
df.groupby('Origin')['Passengers'].sum().sort_values(ascending=False)

,Passengers
Origin,
Delhi,5448
Mumbai,3616
Bengaluru,2725
Srinagar,1957
Kochi,1657
Pune,1564
Ahmedabad,1411
Lucknow,1137
Jaipur,962


In [ ]:
df['Seat_Utilization'].mean()

np.float64(62.66422222222222)

In [ ]:
df['Delay_Category'].value_counts()

,count
Delay_Category,
On Time,70
Minor Delay,63
Major Delay,47


In [ ]:
df.columns.tolist()

['Flight_ID',
 'Flight_Date',
 'Airline',
 'Origin',
 'Destination',
 'Aircraft',
 'Travel_Class',
 'Passengers',
 'Seat_Capacity',
 'Average_Ticket_Price',
 'Delay_Minutes',
 'Flight_Status',
 'Weather',
 'Booking_Channel',
 'Avg_Baggage_Kg',
 'Meal_Preference',
 'Passenger_Satisfaction',
 'Delay_Category',
 'Seat_Utilization']

## 8. DateTime Analysis

In [ ]:
df['Flight_Date'] = pd.to_datetime(df['Flight_Date'])

In [ ]:
df['Flight_Year'] = df['Flight_Date'].dt.year

In [ ]:
df['Flight_Month'] = df['Flight_Date'].dt.month

In [ ]:
df['Flight_Day'] = df['Flight_Date'].dt.day

In [ ]:
df['Flight_Day_Name'] = df['Flight_Date'].dt.day_name()

In [ ]:
df[['Flight_Date', 'Flight_Year', 'Flight_Month', 'Flight_Day', 'Flight_Day_Name']].head(10)

,Flight_Date,Flight_Year,Flight_Month,Flight_Day,Flight_Day_Name
0,2026-03-23,2026,3,23,Monday
1,2026-06-07,2026,6,7,Sunday
2,2026-05-12,2026,5,12,Tuesday
3,2026-05-30,2026,5,30,Saturday
4,2026-02-27,2026,2,27,Friday
5,2026-05-20,2026,5,20,Wednesday
6,2026-06-17,2026,6,17,Wednesday
7,2026-04-28,2026,4,28,Tuesday
8,2026-01-25,2026,1,25,Sunday
9,2026-06-12,2026,6,12,Friday


In [ ]:
df['Flight_Month'].value_counts().sort_index()

,count
Flight_Month,
1,29
2,28
3,33
4,35
5,30
6,25


In [ ]:
df.groupby('Flight_Month').size().sort_values(ascending=False)

,0
Flight_Month,
4,35
3,33
5,30
1,29
2,28
6,25


In [ ]:
df['Flight_Day_Name'].value_counts()

,count
Flight_Day_Name,
Tuesday,38
Saturday,28
Sunday,25
Thursday,24
Monday,23
Friday,23
Wednesday,19


In [ ]:
# Flight status analysis
print("Flights by Status:")
print(df['Flight_Status'].value_counts())

# Average delay by weather condition
print("\nAverage Delay by Weather:")
print(df.groupby('Weather')['Delay_Minutes'].mean().sort_values(ascending=False))

# Passengers by travel class
print("\nPassengers by Travel Class:")
print(df.groupby('Travel_Class')['Passengers'].sum().sort_values(ascending=False))

# Average satisfaction by booking channel
print("\nAverage Satisfaction by Booking Channel:")
print(df.groupby('Booking_Channel')['Passenger_Satisfaction'].mean().sort_values(ascending=False))

# Most common destination
print("\nMost Common Destinations:")
print(df['Destination'].value_counts().head(10))

Flights by Status:
Flight_Status
On Time      114
Delayed       59
Cancelled      7
Name: count, dtype: int64

Average Delay by Weather:
Weather
Rain      22.219512
Storm     20.103448
Cloudy    20.028571
Clear     18.029412
Fog       15.794118
Name: Delay_Minutes, dtype: float64

Passengers by Travel Class:
Travel_Class
Economy            16054
Premium Economy     4574
Business            1471
Name: Passengers, dtype: int64

Average Satisfaction by Booking Channel:
Booking_Channel
Online Travel Portal    3.740000
Mobile App              3.444444
Travel Agency           3.386364
Airline Website         3.365854
Name: Passenger_Satisfaction, dtype: float64

Most Common Destinations:
Destination
Delhi        51
Mumbai       30
Bengaluru    28
Kolkata      20
Chennai      17
Srinagar     13
Hyderabad    13
Kochi         8
Name: count, dtype: int64


## 10. Key Observations

1. Most flights were **On Time (114)**, while **59 flights were Delayed** and **7 flights were Cancelled**.

2. Flights during **Rainy weather conditions** had the highest average delay of approximately **22.22 minutes**.

3. **Economy class** had the highest number of passengers with **16,054 passengers**, followed by Premium Economy and Business class.

4. Passengers who booked through an **Online Travel Portal** had the highest average satisfaction score of **3.74**.

5. **Delhi** was the most common destination with **51 flights**, followed by Mumbai and Bengaluru.

6. Among the weather conditions, **Fog** had the lowest average delay of approximately **15.79 minutes**.

7. The number of cancelled flights was relatively low, with only **7 cancelled flights** in the dataset.

8. The analysis shows that weather conditions, travel class, booking channel, and destination are useful factors for understanding flight operations and passenger experience.

## 11. Conclusion

The Flight Operations dataset was analyzed using Pandas through data inspection, selection, filtering, sorting, grouping, aggregation, transformation, and date-time analysis. The analysis identified patterns in flight delays, passenger distribution, travel classes, booking channels, weather conditions, and destinations. These findings provide useful insights into flight operations and passenger satisfaction.